src/common/jsonb/src/functions.rs

```diff
@@ -0,0 +1,313 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::convert::TryInto;
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+
+use byteorder::BigEndian;
+use byteorder::WriteBytesExt;
+use decimal_rs::Decimal;
+
+use super::constants::*;
+use super::error::*;
+use crate::jentry::JEntry;
+
+// builtin functions for `JSONB` bytes without decode all Values
+
+// build `JSONB` array from items
+// Assuming that the input values is valid JSONB data
+pub fn build_array(values: Vec<Vec<u8>>, buf: &mut Vec<u8>) -> Result<(), Error> {
+    let header = ARRAY_CONTAINER_TAG | values.len() as u32;
+    buf.write_u32::<BigEndian>(header).unwrap();
+
+    let mut jentry_index = 4;
+    // reserve space for jentries
+    buf.resize(4 + values.len() * 4, 0);
+
+    for value in values.iter() {
+        let header = read_u32(value, 0)?;
+        let encoded_jentry = match header & CONTAINER_HEADER_TYPE_MASK {
+            SCALAR_CONTAINER_TAG => {
+                let jentry = &value[4..8];
+                buf.extend_from_slice(&value[8..]);
+                jentry.try_into().unwrap()
+            }
+            ARRAY_CONTAINER_TAG | OBJECT_CONTAINER_TAG => {
+                buf.extend_from_slice(value);
+                (CONTAINER_TAG | value.len() as u32).to_be_bytes()
+            }
+            _ => return Err(Error::InvalidJsonbHeader),
+        };
+        for (i, b) in encoded_jentry.iter().enumerate() {
+            buf[jentry_index + i] = *b;
+        }
+        jentry_index += 4;
+    }
+    Ok(())
+}
+
+// `JSONB` values supports partial decode for comparison,
+// if the values are found to be unequal,
+// the result will be returned immediately
```

Comment:
Maybe the comment `// Null > Container(Array > Object) > String > Number > Boolean` should on here? Not on `compare_scalar`?

Reply:
those are `JEntry` types, I will add some comments for `fn compare`